### PR TITLE
Assembly: BOM follow up

### PR DESCRIPTION
**Pull request by PaddleStroke**
- Fixes some comments from @wwmayer 
- Make auto-generated columns bold to differentiate them from custom columns
- Add a help button with details
- Rename 'Detail parts' to 'Parts children' as I think it's clearer.

image

### CODE REVIEW

File: src/Mod/Assembly/CommandCreateBom.py [hunk 1]
```diff
@@ -289,6 +324,71 @@ def eventFilter(self, watched, event):
...
+                "Only parts : If checked, only Part containers and sub-assemblies will be added to the bill of materials. Solids like PartDesign Bodies, fasteners or Part workbench primitives will be ignored.",
+            )
+            + "\n"
+        )
+        columns_title = QtWidgets.QLabel("<b>" + translate("Assembly", "Columns:") + "</b>")
+        columns_text = QtWidgets.QLabel(
+            " - "
+            + translate(
+                "Assembly",
+                "Bold columns :  (Index, Quantity, Name...) are populated automatically. Any modification you make will be overriden. These columns cannot be renamed.",
```
Comment (chennes) [inline src/Mod/Assembly/CommandCreateBom.py:362]:
```suggestion
                "Bold columns :  (Index, Quantity, Name...) are populated automatically. Any modification you make will be overridden. These columns cannot be renamed.",
```

**Comment by pierreporte** [comment]
The BOM type names are still harder to understand than first level BOM, multi-level BOM, and components list, as described in my comment on the previous PR: https://github.com/FreeCAD/FreeCAD/pull/14198#issuecomment-2154593346. Better text reduces the need of a help dialog (in fact, tooltips would be sufficient). BTW the only parts option is said to count assemblies which is misleading.

The column list can be better design, in a way that removes the need of a help dialog. The bold style meaning is not obvious: I propose an italic text after the column name, in parentheses:

> File name *(auto)*

I propose to replace the add column button by a dropdown list, with all available options (with the auto indicator on relevant lines) and a “custom…” line that gives the focus to an empty text field in the list as is currently work. If an automatic column is removed, the user wouldn’t need to know its exact name. To remove columns, I propose to keep the del key working as it is and adding a remove column button with the same size on the left of the add button.

I believe that these changes will allow to remove the help dialog.

File: src/Mod/Assembly/CommandCreateBom.py (hunk 1 above)
Comment (PaddleStroke) [reply src/Mod/Assembly/CommandCreateBom.py]: Thanks

**Comment by PaddleStroke** [comment]
> The BOM type names are still harder to understand than first level BOM, multi-level BOM, and components list, as described in my comment on the previous PR: https://github.com/FreeCAD/FreeCAD/pull/14198#issuecomment-2154593346. Better text reduces the need of a help dialog (in fact, tooltips would be sufficient). BTW the only parts option is said to count assemblies which is misleading.

@pierreporte I have seen your comment, and yesterday I was looking at it over and over. My problem is a translation problem. I failed to translate the windchill help into freecad concepts. I think the concept you are talking about are not a match to the options I created.

> single level BOM: list of direct children of the root assembly;
> multi-level BOM: list of direct children of the root assembly, children sub-assemblies are themselves followed by their multi-level BOM (think a tree view) with an explicit hierarchy;
> multi-level component list: flattened multi-level BOM, duplicates are merged and their quantity added.

For instance you mention that multi-level component is flattened. I don't think any of the current options does that.

I am not opposed to use the options you mentioned. But I cannot just change the strings, the tool needs to behave as user would expect from the names. So I need more input : 
- Which softwares are using these concepts? Just to make sure it's industry standard.
- Can you please detail exactly how each option should work in freecad concepts? Please give me some BOM simple examples for each option to make sure it's clear.

**Comment by PaddleStroke** [comment]
Regarding the columns :  
- Adding (Auto) to the name is acceptable.
- Using a button with a menu is also a good idea.

Rather than a remove column, I wonder if I can add 'x' buttons on each item that would remove the column.

**Comment by PaddleStroke** [comment]
Regarding the 'Only part' option, what name do you think would be better?
This option is rather 'List Part::Feature' but I think users are not going to understand. I could use 'List Bodies' which people would understand better, but that does not mention other kind of Part::Features.

**Comment by pierreporte** [comment]
@PaddleStroke The different BOM types are used in all PDM/PLM solutions: Windchill, 3DExperience, Smarteam, … This is one of the most basic feature of this kind of system. The names can vary a little bit but it’s always the same principle.

Below is an example of several BOMs generated from a product structure (invented just for you but it’s realistic). The structure is what would be the FreeCAD design tree, limited to parts and assemblies (parts are lines with no children), with instance numbers. Expanded BOMs don’t group together parts that are identical while consolidated BOMs do. Multi-level BOMs give the whole structure while single level BOMs give only the direct children of the root. The components list is a flattened and consolidated multi-level BOM.

Regarding the “only parts” option, it would be best to settle the Part/Body problem first. It’s truly needed.

---------------

### Product Structure

* EQUIPPED MANIFOLD ASSEMBLY
  * BARE MANIFOLD<1>
  * HEX SCREW M6x60<1>
  * HEX SCREW M6x60<2>
  * HEX SCREW M6x60<3>
  * HEX SCREW M6x60<4>
  * JIC DASH 10 UNION<1>
  * JIC DASH 10 UNION<2>
  * JIC DASH 6 UNION<1>
  * JIC DASH 6 UNION<2>
  * JIC DASH 6 UNION<3>
  * JIC DASH 6 UNION<4>
  * O-RING 15x1.78 NBR<1>
  * O-RING 15x1.78 NBR<2>
  * O-RING 10x1.78 NBR<1>
  * O-RING 10x1.78 NBR<2>
  * O-RING 10x1.78 NBR<3>
  * O-RING 10x1.78 NBR<4>
  * LEE PLUG<1>
  * LEE PLUG<2>
  * LEE PLUG<3>
  * LEE CHECK VALVE<1>
  * 100 MICRON FILTER<1>
  * FLOWMETER<1>
    * PROPELLER CARTRIDGE<1>
    * HALL EFFECT SENSOR<1>
    * O-RING 15x1.78 NBR<1>
    * HEX SOCKET SCREW M3x8<1>
    * HEX SOCKET SCREW M3x8<2>
    * HEX SOCKET SCREW M3x8<3>
    * HEX SOCKET SCREW M3x8<4>
  * EQUIPPED PT100B TEMPERATURE SENSOR<1>
    * PT100B TEMPERATURE SENSOR<1>
    * O-RING 10x1.78 NBR<1>
    * 4-PIN CONNECTOR<1>
  * EQUIPPED PT100B TEMPERATURE SENSOR<2>
    * PT100B TEMPERATURE SENSOR<1>
    * O-RING 10x1.78 NBR<1>
    * 4-PIN CONNECTOR<1>
  * LABEL<1>
  * CAPTIVE DRAINAGE PLUG<1>
    * BARE DRAINAGE PLUG<1>
    * O-RING 15x1.78 NBR<1>
    * HEX SOCKET SCREW M3x8<1>
    * HEX SOCKET SCREW M3x8<2>
    * EQUIPPED ROPE<1>
      * NYLON ROPE 2MM<1>
      * CLIMPLED CLIP<1>
      * CRIMPLED CLIP<2>
      * ATTACH PLATE<1>
      * ATTACH PLATE<2>

### Multi-level expanded BOM

Identical to product structure, with a similar format to the following.

### Multi-level consolidated BOM

| Level  | Component                          | Qty |
| ------ | ---------------------------------- | --- |
| 1      | BARE MANIFOLD                      | 1   |
| 2      | HEX SCREW M6x60                    | 4   |
| 3      | JIC DASH 10 UNION                  | 2   |
| 4      | JIC DASH 6 UNION                   | 4   |
| 5      | O-RING 15x1.78 NBR                 | 2   |
| 6      | O-RING 10x1.78 NBR                 | 4   |
| 7      | LEE PLUG                           | 3   |
| 8      | LEE CHECK VALVE                    | 1   |
| 9      | 100 MICRON FILTER                  | 1   |
| 10     | FLOWMETER                          | 1   |
| 10.1   | PROPELLER CARTRIDGE                | 1   |
| 10.2   | HALL EFFECT SENSOR                 | 1   |
| 10.3   | O-RING 15x1.78 NBR                 | 1   |
| 10.4   | HEX SOCKET SCREW M3x8              | 4   |
| 11     | EQUIPPED PT100B TEMPERATURE SENSOR | 2   |
| 11.1   | PT100B TEMPERATURE SENSOR          | 1   |
| 11.2   | O-RING 10x1.78 NBR                 | 1   |
| 11.3   | 4-PIN CONNECTOR                    | 1   |
| 12     | LABEL                              | 1   |
| 13     | CAPTIVE DRAINAGE PLUG              | 1   |
| 13.1   | BARE DRAINAGE PLUG                 | 1   |
| 13.2   | O-RING 15x1.78 NBR                 | 1   |
| 13.3   | HEX SOCKET SCREW M3x8              | 2   |
| 13.4   | EQUIPPED ROPE                      | 1   |
| 13.4.1 | NYLON ROPE 2MM                     | 1   |
| 13.4.2 | CLIMPLED CLIP                      | 2   |
| 13.4.3 | ATTACH PLATE                       | 2   |

### Single level BOM

| Level  | Component                          | Qty |
| ------ | ---------------------------------- | --- |
| 1      | BARE MANIFOLD                      | 1   |
| 2      | HEX SCREW M6x60                    | 1   |
| 3      | HEX SCREW M6x60                    | 1   |
| 4      | HEX SCREW M6x60                    | 1   |
| 5      | HEX SCREW M6x60                    | 1   |
| 6      | JIC DASH 10 UNION                  | 1   |
| 7      | JIC DASH 10 UNION                  | 1   |
| 8      | JIC DASH 6 UNION                   | 1   |
| 9      | JIC DASH 6 UNION                   | 1   |
| 10     | JIC DASH 6 UNION                   | 1   |
| 11     | JIC DASH 6 UNION                   | 1   |
| 12     | O-RING 15x1.78 NBR                 | 1   |
| 13     | O-RING 15x1.78 NBR                 | 1   |
| 14     | O-RING 10x1.78 NBR                 | 1   |
| 15     | O-RING 10x1.78 NBR                 | 1   |
| 16     | O-RING 10x1.78 NBR                 | 1   |
| 17     | O-RING 10x1.78 NBR                 | 1   |
| 18     | LEE PLUG                           | 1   |
| 19     | LEE PLUG                           | 1   |
| 20     | LEE PLUG                           | 1   |
| 21     | LEE CHECK VALVE                    | 1   |
| 22     | 100 MICRON FILTER                  | 1   |
| 23     | FLOWMETER                          | 1   |
| 24     | EQUIPPED PT100B TEMPERATURE SENSOR | 1   |
| 25     | EQUIPPED PT100B TEMPERATURE SENSOR | 1   |
| 26     | LABEL                              | 1   |
| 27     | CAPTIVE DRAINAGE PLUG              | 1   |

### Single level consolidated BOM

| Level  | Component                          | Qty |
| ------ | ---------------------------------- | --- |
| 1      | BARE MANIFOLD                      | 1   |
| 2      | HEX SCREW M6x60                    | 4   |
| 3      | JIC DASH 10 UNION                  | 2   |
| 4      | JIC DASH 6 UNION                   | 4   |
| 5      | O-RING 15x1.78 NBR                 | 2   |
| 6      | O-RING 10x1.78 NBR                 | 4   |
| 7      | LEE PLUG                           | 3   |
| 8      | LEE CHECK VALVE                    | 1   |
| 9      | 100 MICRON FILTER                  | 1   |
| 10     | FLOWMETER                          | 1   |
| 11     | EQUIPPED PT100B TEMPERATURE SENSOR | 2   |
| 12     | LABEL                              | 1   |
| 13     | CAPTIVE DRAINAGE PLUG              | 1   |

### Components list

| Component                 | Qty |
| ------------------------- | --- |
| BARE MANIFOLD             | 1   |
| HEX SCREW M6x60           | 4   |
| JIC DASH 10 UNION         | 2   |
| JIC DASH 6 UNION          | 4   |
| O-RING 15x1.78 NBR        | 4   |
| O-RING 10x1.78 NBR        | 6   |
| LEE PLUG                  | 3   |
| LEE CHECK VALVE           | 1   |
| 100 MICRON FILTER         | 1   |
| PROPELLER CARTRIDGE       | 1   |
| HALL EFFECT SENSOR        | 1   |
| HEX SOCKET SCREW M3x8     | 5   |
| PT100B TEMPERATURE SENSOR | 2   |
| 4-PIN CONNECTOR           | 2   |
| LABEL                     | 1   |
| BARE DRAINAGE PLUG        | 1   |
| NYLON ROPE 2MM            | 1   |
| CLIMPLED CLIP             | 2   |
| ATTACH PLATE              | 2   |

**Comment by pierreporte** [comment]
@PaddleStroke did you find my example useful?

**Comment by PaddleStroke** [comment]
@pierreporte thanks for the example. I need to check it out, but I have few bugs/crashes to fix first before going back to BOM tool.

**Comment by pierreporte** [comment]
@PaddleStroke Do you plan a second follow-up?

**Comment by PaddleStroke** [comment]
Yes, I will implement as you asked but after 1.0 because this requires quite a lot of changes and it's actually a feature request.

**Comment by Demolux** [comment]
I would like to help with BOM feature expansion and would find options for different BOM structures very useful too.
I did focus my first dev approach on the auto generation feature because I wanted to have additional information added dynamically, such as Price, Material, Source, a Dimension, Notes and so on.

So what about using the properties of an object for an auto fill feature for columns? I made a small implementation, see here:

tl dr;
If you add " (auto)" as suffix to your column name, FreeCAD will search for a property with that name and display its value in the according cell of the object and column.

9EQKTJl

Idk which is the right place do discuss this, here or the forums. I created a more detailed forum post yesterday

https://forum.freecad.org/viewtopic.php?t=93137&sid=ac528ccaf16849db78069c1575d6d78c

@PaddleStroke hinted me to discuss it here with @pierreporte

Can someone also point me to discussions about the Part/Body Problem especially regarding possible future changes? I totally agree that will directly affect BOM generation and there should be a somehow solid structural foundation first

**Comment by pierreporte** [comment]
@Demolux The problem with parts and bodies is one of those things that originates from the lack of design of the whole program. Bodies were introduced first and remained the default container for parts made with PartDesign, thus the most common, even after the Part container was introduced. So much that the new Assembly workbench allows users to directly insert bodies. In the professional CAD world, it is common to allow parts to have several bodies, but they never allow standalone bodies. In my opinion key question is: should FreeCAD allow multi-body parts? If not, the Part container is useless. If yes, then there must be a container for them: this is Part. There have been a discussion about this on Discord within the Design Working Group (the CAD Working Group didn't exist at the time) after which we agreed on the need to allow mult-body parts, while discussing about how to manage bodies after the Part/PartDesign merge. The other big question is, that considering the history of FreeCAD, should we allow standalone bodies in assemblies? I do think that the answer should be no (but it would now be a breaking change) because in CAD bodies are never considered standalone objects and doing so entertains the confusion. The problem partially lies in UI/UX because marking Part container is harder while clicking on the start page tile instantly creates a Body.

**Comment by Demolux** [comment]
@pierreporte Thanks for the sum-up and explanation. 

____comparing FreeCAD to Creo and Fusion360____
I think though the answer is not that clear as there is at least Fusion360 which allows single bodies to exist. It's not aimed strictly for professional use though, such as Creo which in turn does not allow that. FreeCAD also is not aimed strictly for professional use.

Like you introduced, allowing standalone bodies indeed simplifies on-boarding for new users, rapid prototyping or occasional use from hobbyists. As of FreeCAD 1.0 that is taken the furthest of all CAD programs I know by the "allow compounds" feature for a body, which can have multiple, seperate geometries that way. Even Fusion does not allow that, here a split body automatically becomes seperate bodies. Creo allows it though - but always uses part containers. Fusion also has part containers (called components), but does not auto-use it when you create a body. So regarding that and the CAD world I think we have quite some flexibility and the current solution probably suits FreeCAD best regarding its user group.

When it comes to a BOM though its a different story and the user group probably changes more into the professional use. Fusion and Creo both do not list bodies in the BOM (unless you take action to change that for specific bodies). They also only allow part containers in assemblies. But who says that's the only way - why not make it optional to also list bodies generally.

____possible features for FreeCAD regarding BOM generation____
Feature:
The current BOM generation allows to check "Only parts" in the dialogue but that also excludes any other parts you may want to count such as screws, draftobjects such as arrays and other things you may want to count. So why not go the other way and make a checkbox "Exclude bodies" instead, or a list where you can exclude other objecttypes as well. This would make BOm generation very flexible and customizable.

Feature:
I also would like to have dynamically filled custom columns in one way or another (e. g. the column-name/property matching approach shown above).

Feature:
UI/UX improvements but thats something I did not look into at all yet, but its "feelable" there is potential

____questions____
I would like to start going a little deeper into development but A don't want to work against or parallel to someone and B don't want to work completly off-road noticing that there is some major overhaul affecting all that a few months later. Is there some group or space where the BOM development is discussed? Or is it more like "implement it and make a showcasing PR"?

**Comment by furgo16** [comment]
@Demolux for better visibility, please consider opening a new issue for discussion. Adding comments on a PR closed 6 months ago will make the discussion effectively visible only to those that subscribed to that PR back in the day and still have notifications enabled. You already posted on the forum, which is great, but if you want to continue the conversation on GitHub with a wider audience, a new issue will get more exposure and will be easier to find and reference. I saw your other PR: thanks for your work on the BOM feature!
